src/daemon/daemon.cpp

```diff
@@ -666,7 +671,19 @@ try // clang-format on
     if (metrics_opt_in.opt_in_status == OptInStatus::ACCEPTED)
         metrics_provider.send_metrics();
 
-    auto name = name_from(request, *config->name_generator, vm_instances);
+    auto mem_disk_name_and_errors = validate_create_arguments(request);
+    auto mem_size = std::get<0>(mem_disk_name_and_errors);      // use structured bindings instead in C++17
+    auto disk_space = std::get<1>(mem_disk_name_and_errors);    // use structured bindings instead in C++17
+    auto instance_name = std::get<2>(mem_disk_name_and_errors); // use structured bindings instead in C++17
+    auto option_errors = std::get<3>(mem_disk_name_and_errors); // use structured bindings instead in C++17
```

Comment:
I dislike this because there's no clear connection between 0 and mem_size, 1 and disk_space, etc. Sure it works, but when we add an entry in the tuple, or swap entries with same type, it's easy to refer to the wrong index and create a bug, yet the code will still compile.

A quick struct will give names to these entries, establishing a stronger key/value connection. e.g. VirtualMachineDescription

Note, you could correctly point out that we do something very similar in the return statements of validate_create_arguments, to_machine_desc, etc - where we use brace initializers to populate a struct - as again there's no clear key/value connection. I'm not a giant fan of that style either, but at least it is localised to the creation of the struct, and not the consumer, so the person who changes it will (I think) be more aware of the consequences.

Reply:
Taking your points
1. as I said, swapping entries of same type will still compile, but create bug. Public/private is irrelevant to that argument, and the greatest strength structs have.
2. I'm only taking this case, not every return. Named returns would be ideal, closest is a struct IMO. 
3. same problem, just replace indexes with key names.
4. I'm not saying it's ideal, but least worst option IMO
5. not just public IMO
6. readability isn't my priority here, avoiding bugs is - especially those possibly introduced later by a different author.
7. it's a struct just to pass a list of values with labels. My objection is purely down to the lack of labels with your approach here.
8. I dunno, ValidatedVMOptions?

Sorry, my opinion hasn't changed. Note I'm not making a big deal here, it's just code after all :) 

I'll let @townsend2010 have his say.

Reply:
I do feel like the proposed approach is more prone to bugs accidentally introduced if some other maintainer or contributor down the road added another check here, like for example, validating the number of cores/CPU's as well.  What I'm not a fan of, is that the consumer **has** to know the exact order of the elements that are returned from `validate_create_arguments`.  With a struct, it only needs to know the struct's member name and the rest still are valid, even if a contributor puts a new member variable anywhere in the struct.

But as @gerboland said, it's just code and it works now, so I wouldn't block on this either:)

Reply:
Thanks for the comments. No big deal here either! Just debating an interesting issue with implications in future coding practices :) I actually agree with you to a large extent, but I don't quite follow what you say in 1 and 6.

Suppose I used a `struct Foo`, with ctor `Foo(mem, disk, name, errors)`. The function would say `return Foo{a,b,c,d};` right? Aren't argument swapping bugs as likely? The way I see it, a change in the producer requires a change in the consumer in both cases. The compiler will do the same. What changes is where those responsibilities lie, which is why I mentioned public/private.

A struct-based approach encapsulates the consumer, preventing impact on the function's client(s). But if the function has a single client and is only there to offload code into the anonymous namespace, the benefit is small IMO. When considering also drawbacks (struct proliferation, *a la carte* types representing no clear concept), I place the ideal compromise a bit differently (barring named returns).

Anyway... I just saw @townsend2010's comment and will modify the code to use the struct.

Reply:
I just committed changes with the ordering now encapsulated in a struct. Does that address the concerns?